llvm/lib/IR/RepoTicket.cpp

```diff
@@ -163,91 +143,89 @@ updateDigestUseCallDependencies(const GlobalObject *GO, MD5 &GOHash,
+  Changed = Changed || std::get<2>(State);
+
+  // Update the GO's hash using its contributions and dependencies.
+  auto &GOContributions = std::get<0>(State);
+  auto &GODependencies = std::get<1>(State);
+  for (const GlobalObject *const G : llvm::concat<const GlobalObject *const>(
+           make_range(GOContributions.begin(), GOContributions.end()),
+           make_range(GODependencies.begin(), GODependencies.end()))) {
+    const llvm::Function *const Fn = dyn_cast<const llvm::Function>(G);
     // if function will not be inlined, skip it
     if (Fn && Fn->hasFnAttribute(Attribute::NoInline))
       continue;
-    Changed = updateDigestUseCallDependencies(D, GOHash, Visited, GOIMap,
-                                              UpdateDigestAndGetDependencies) ||
+    Changed = updateDigestUseDependenciesAndContributions(
+                  G, GOHash, Visited, GOIMap, UpdateDigestAndGetGOInfo) ||
               Changed;
   }
+
   return Changed;
 }
 
+// Calculate the GOs' initial digest, dependencies, contributions and GONumber.
+std::pair<GOInfoMap, GONumber> caculateGONumAndGOIMap(Module &M) {
+  GOInfoMap GOIMap;
+  GONumber GONum;
+
+  for (auto &GV : M.globals()) {
+    if (GV.isDeclaration())
+      continue;
+    calculateInitialDigestAndDependenciesAndContributions(&GV, GOIMap);
+    ++GONum.VarNum;
+  }
+  for (auto &Fn : M.functions()) {
+    if (Fn.isDeclaration())
+      continue;
+    calculateInitialDigestAndDependenciesAndContributions(&Fn, GOIMap);
+    ++GONum.FuncNum;
+  }
+  return std::make_pair(std::move(GOIMap), std::move(GONum));
+}
+
 // Calculate the final GO hash by adding the initial hashes of its dependents
 // and the contributions and create the ticket metadata for GOs.
 std::tuple<bool, unsigned, unsigned> generateTicketMDs(Module &M) {
   bool Changed = false;
   GONumber GONum;
   GOStateMap Visited;
+  Visited.reserve(M.size() + M.getGlobalList().size());
   GOInfoMap GOIMap;
-  GVInfoMap GVIMap;
 
-  for (auto &GO : M.global_objects()) {
-    if (GO.isDeclaration())
-      continue;
-    Visited.clear();
-    MD5 Hash = MD5();
-    auto Helper = [&GONum](const GlobalObject *GO, MD5 &GOHash,
-                           GOInfoMap &GOIMap) {
-      return updateDigestGONumAndGetDependencies(GO, GOHash, GOIMap, GONum);
-    };
-    updateDigestUseCallDependencies(&GO, Hash, Visited, GOIMap, Helper);
-    for (const auto GV : GOIMap[&GO].Contributions) {
-      // Record GO's possible contributions.
-      GVIMap[GV].emplace_back(&GO);
-    }
-
-    MD5::MD5Result Digest;
-    Hash.final(Digest);
-    set(&GO, Digest);
-    Changed = true;
-  }
-
-  if (GVIMap.empty()) {
-    return std::make_tuple(Changed, GONum.VarNum, GONum.FuncNum);
-  }
-
-  // Update the GV's digest using the contributions.
-  for (auto &Entry : GVIMap) {
-    if (Entry.first->isDeclaration())
-      continue;
+  // Calculate the GOs' initial digest, dependencies, contributions.
```

Comment:
The digest, dependencies and contributions should be encapsulated inside GOIMap and GOInfo so I don't think this comment should refer to them individually.

Reply:
Fixed, thanks.